File: worker/inventory.py
"""Versioned candidate inventory checksum; consistency, never attestation."""
import hashlib
import json
import re

FORMAT = 'iim.extensions.v2'


def clean_entries(entries):
    if not isinstance(entries, list) or len(entries) > 250:
        raise ValueError('Expected up to 250 extension entries')
    clean = []
    for row in entries:
        if not isinstance(row, dict) or not isinstance(row.get('name'), str) or not isinstance(row.get('enabled'), bool):
            raise ValueError('Each entry requires a name and enabled boolean')
        clean.append({'name': row['name'][:100], 'enabled': row['enabled']})
    return clean


def checksum(clean, format=FORMAT):
    if format == FORMAT:
        # ASCII-escaped rows, sorted by ASCII bytes, including the boolean tie-break.
        rows = sorted(json.dumps([r['name'], r['enabled']], ensure_ascii=True, separators=(',', ':')) for r in clean)
        encoded = '[' + ','.join(rows) + ']'
    elif format in (None, 'iim.extensions.v1'):
        rows = sorted([[r['name'], r['enabled']] for r in clean], key=lambda r: r[0])
        encoded = json.dumps(rows, separators=(',', ':'))
    else:
        raise ValueError('Unsupported extension inventory format; export it again with the current companion')
    return hashlib.sha256(encoded.encode('utf-8')).hexdigest()


def validate_inventory(entries, digest=None):
    format = None
    if isinstance(entries, dict):
        format = entries.get('format')
        digest = entries.get('digest', digest)
        entries = entries.get('extensions')
    if format not in (None, 'iim.extensions.v1', FORMAT):
        raise ValueError('Unsupported extension inventory format; export it again with the current companion')
    clean = clean_entries(entries)
    if format == FORMAT and digest is None:
        raise ValueError('Version 2 inventory requires a checksum')
    if digest is not None:
        if not isinstance(digest, str) or not re.fullmatch('[0-9a-fA-F]{64}', digest):
            raise ValueError('Expected a 64-character SHA-256 checksum')
        if checksum(clean, format) != digest.lower():
            raise ValueError('Inventory checksum does not match the file contents; export the inventory again. A checksum is not browser authentication.')
    return clean, digest is not None

File: companion/inventory.js
// v2: truncate to 100 Unicode code points, ASCII-escape JSON rows, sort ASCII rows.
// A checksum confirms consistency of editable content; it is not browser attestation.
export async function exportInventory(inventory){
  if(!Array.isArray(inventory)||inventory.length>250)throw Error('Expected up to 250 extensions');
  const clean=inventory.map(row=>{
    if(typeof row.name!=='string'||typeof row.enabled!=='boolean')throw Error('Invalid extension entry');
    return {name:Array.from(row.name).slice(0,100).join(''),enabled:row.enabled};
  });
  const rows=clean.map(row=>JSON.stringify([row.name,row.enabled])
    .replace(/[\u007f-\uffff]/g,c=>'\\u'+c.charCodeAt(0).toString(16).padStart(4,'0'))).sort();
  const bytes=new TextEncoder().encode('['+rows.join(',')+']');
  const hash=await crypto.subtle.digest('SHA-256',bytes);
  const digest=Array.from(new Uint8Array(hash),b=>b.toString(16).padStart(2,'0')).join('');
  return {format:'iim.extensions.v2',exported_at:new Date().toISOString(),self_reported:true,digest,extensions:clean};
}
